### PR TITLE
Remove Popup and ContentPage from LogicalChildren when Popup is closed on iOS

### LINKED ISSUE #2149: [BUG] iOS: Popup Remains in Page's LogicalChildren after Closing
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

When we create a popup on iOS we add two Elements to the Page's LogicalChildren: the Popup, and the ContentPage housing the Content.

When the popup is closed, the former is removed from the Page's Logical Children but the latter is not. This causes the popup (and its ViewModel) to actually remain in the heap until the page is popped and the reference (the Page's logical children) is cleared out.

### Expected Behavior

The Popup and ContentPage should be removed from the Logical Children.

We can't really rely on popping the page to do this for us because the page in question could be the home page of an application which may not ever be popped.

### Steps To Reproduce

1. Create a page with two buttons
2. One button, Show Popup, should create a new popup and call App.Current.MainPage.ShowPopup
3. The other, Collect Garbage, should force Garbage Collection
4. Run this on iOS and tap ShowPopup a few times
5. Tap Collect Garbage (a few times for good measure...)
6. Get a GC dump

When looking at the GC dump you'll see the popups exist in memory. _In my reproduction sample I'm checking WeakReference.IsAlive, but this was after checking GC dumps manually._

### Link to public reproduction project repository

https://github.com/NeilMalcolm/PopupInMemoryIssue

In my reproduction project, I've added a static boolean which can be toggled to enable the 'fix' for subsequent popups which allows these to be garbage collected. The fix was made a subclass of the PopupHandler, see here https://github.com/NeilMalcolm/PopupInMemoryIssue/blob/main/src/MauiPopupInMemory/Platforms/iOS/MyPopupHandler.cs

My fix is related to issue https://github.com/CommunityToolkit/Maui/issues/1676. When the page is popped, the popup is able to be garbage collected, but as long as the page remains in the navigation stack it won't because the ContentPage remains a Logical Child of the Page.

### Environment

```markdown
- .NET MAUI CommunityToolkit:9.0.3
- OS:iOS
- .NET MAUI: 8.0.80
```

### Anything else?

Thank you all for maintaining this project, it's so important for MAUI.

### PR DESCRIPTION
**Pull request by cat0363**
This PR resolves the issue where Popup and Content Page are not removed from LogicalChildren after closing Popup on iOS.

 ### Description of Change ###

To accomplish this, first add MapOnClosed to the ControlPopUpCommandMapper.

[src\CommunityToolkit.Maui\HandlerImplementation\Popup\Popup.shared.cs]

```
	public static CommandMapper<IPopup, PopupHandler> ControlPopUpCommandMapper = new(PopupHandler.PopUpCommandMapper)
	{
#if IOS || MACCATALYST
		[nameof(IPopup.OnOpened)] = MapOnOpened,
		[nameof(IPopup.OnClosed)] = MapOnClosed
#endif
	};
```

Next, within the MapOnClosed method, remove Popup and ContentPage from LogicalChildren.

[src\CommunityToolkit.Maui\HandlerImplementation\Popup\Popup.macios.cs]

```
	public static void MapOnClosed(PopupHandler handler, IPopup view, object? result)
	{
		var parent = view.Parent as Element;
		if (parent is not null)
		{
			var contentPage = parent.LogicalChildrenInternal
				.Where(
					x => x is ContentPage
				)
				.LastOrDefault();

			var popup = parent.LogicalChildrenInternal
				.Where(
					x => x is Popup
				)
				.LastOrDefault();

			if (contentPage is not null)
			{
				parent.RemoveLogicalChild(contentPage);
			}
			if (popup is not null)
			{
				parent.RemoveLogicalChild(popup);
			}
		}
	}
```

Since Popup and ContentPage are stacked in LogicalChildren, remove the Popup and ContentPage added to LogicalChildren when the Popup was displayed last.

This will remove Popups and ContentPages from LogicalChildren in the order that Popups are closed.

 

 ### Linked Issues ###
 

 - Fixes #2149

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls 

 ### Additional information ###

 

Below are the verification results.
For verification, I used the reproduction code of Issue #2149.

https://github.com/user-attachments/assets/eca19627-3654-476e-8b60-d21195c4aece

You can see that Popup and ContentPage are removed from LogicalChildren.

Next, I verified a pattern that displays popups hierarchically.

https://github.com/user-attachments/assets/ee4f690f-bd05-438e-9f4f-0d08a9783d4c

Similarly, you can see that Popup and ContentPage have been removed from LogicalChildren.

### CODE REVIEW

File: src/CommunityToolkit.Maui/HandlerImplementation/Popup/Popup.macios.cs [hunk 1]
```diff
@@ -35,4 +35,38 @@ static PageHandler CreatePageHandler(IPopup virtualView)
...
+	public static void MapOnClosed(PopupHandler handler, IPopup view, object? result)
+	{
+		var parent = view.Parent as Element;
+		if (parent is not null)
+		{
+			var contentPage = parent.LogicalChildrenInternal
+				.Where(
+					x => x is ContentPage
+				)
+				.LastOrDefault();
```
Comment (NeilMalcolm) [inline src/CommunityToolkit.Maui/HandlerImplementation/Popup/Popup.macios.cs:54]: Any risk we'll end up removing something that isn't a popup's ContentPage here? 

Can we get a reference to the specific ContentPage using handler.PlatformView.Control?

Comment (cat0363) [reply src/CommunityToolkit.Maui/HandlerImplementation/Popup/Popup.macios.cs:54]: @NeilMalcolm , Thank you for your review.
In cases where multiple Popups are displayed, a problem may arise if the Popups are not closed in the reverse order of the display order. I'll look into another fix.

Comment (cat0363) [reply src/CommunityToolkit.Maui/HandlerImplementation/Popup/Popup.macios.cs]: @NeilMalcolm , I have made the following changes.
With this method, there is no need to consider the display order of Popups.

```
	public static void MapOnClosed(PopupHandler handler, IPopup view, object? result)
	{
		var parent = view.Parent as Element;
		if (parent is not null)
		{
			if (handler.VirtualView is Popup popup)
			{
				if (popup.Content is not null)
				{
					if (popup.Content.Parent is ContentPage contentPage)
					{
						parent.RemoveLogicalChild(contentPage);
					}
				}
				parent.RemoveLogicalChild(popup);
			}
		}
	}
```

**Review by bijington — APPROVED** [review]
Thanks @cat0363 this is great work as always

**Comment by cat0363** [comment]
@bijington , I found that the following code causes a problem in which Popup does not close when calling the CloseAsync or Close method of the Popup class.

[src\CommunityToolkit.Maui\HandlerImplementation\Popup\Popup.shared.cs]

```
	public static CommandMapper<IPopup, PopupHandler> ControlPopUpCommandMapper = new(PopupHandler.PopUpCommandMapper)
	{
#if IOS || MACCATALYST
		[nameof(IPopup.OnOpened)] = MapOnOpened,
		[nameof(IPopup.OnClosed)] = MapOnClosed        // <= here
#endif
	};
```

The above code prevents the following methods from being called.

[src\CommunityToolkit.Maui.Core\Handlers\Popup\PopupHandler.macios.cs]

```
	public static async void MapOnClosed(PopupHandler handler, IPopup view, object? result)
	{
		var presentationController = handler.PlatformView.PresentationController;
		if (presentationController?.PresentedViewController is UIViewController presentationViewController)
		{
			await presentationViewController.DismissViewControllerAsync(true);
		}

		view.HandlerCompleteTCS.TrySetResult();

		handler.DisconnectHandler(handler.PlatformView);
	}
```

I'm currently looking into how to fix it.

**Comment by cat0363** [comment]
@bijington , I have created a new PR #2202.
I apologize that this PR affected the closing of Pop Up by calling the Close or CloseAsync methods.

**Pull request merged**